### PR TITLE
Fix issue 20810: Freshly built DMD fails to read config file 

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
According to MSDN docs the return value of GetFullPathName does *not* count the terminating null character on a successful call.

See here: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamew

> Note  Although the return value in this case is a length that includes the terminating null character, the return value on success does not include the terminating null character in the count.

I'm not sure how this was never caught before though. But on my Win10 machine `master` fails.

Fixes https://issues.dlang.org/show_bug.cgi?id=20810

### CODE REVIEW

File: src/dmd/root/filename.d [hunk 1]
```diff
@@ -1005,10 +1005,13 @@ nothrow:
                 auto fullPath = (cast(wchar*) mem.xmalloc_noscan((fullPathLength + 1) * wchar.sizeof))[0 .. fullPathLength + 1];
                 scope(exit) mem.xfree(fullPath.ptr);
 
+                import core.stdc.stdio;
```
Comment (AndrejMitrovic) [inline src/dmd/root/filename.d:1008]: remove this

**Comment by Geod24** [comment]
Needs a test tho

**Comment by AndrejMitrovic** [comment]
I'll see if I can add a test-case. It's coming from the call to `usage()`, so calling DMD with no params or probably `--help`. I'm guessing there's no such test in the test-suite.

**Comment by AndrejMitrovic** [comment]
I added a test, but I haven't been able to unittest on Windows locally yet. There's some issue with the `build.d` script:

```
rdmd build.d unittest

...

 dmd.obj : error LNK2001: unresolved external symbol _D3dmd7backend8bcomplex12__ModuleInfoZ
C:\dev\dmd\generated\windows\release\64\dmd.exe-unittest : fatal error LNK1120: 1 unresolved externals
```

I'll try and sort this out soon.

**Comment by MoonlightSentinel** [comment]
> I added a test, but I haven't been able to unittest on Windows locally yet. There's some issue with the `build.d` script:

See #11112

**Comment by AndrejMitrovic** [comment]
It looks like the test-suite fails with the change. So it's not the appropriate fix.

I wonder if the bug is related to any recent Windows updates. I'm running 1909. But the only changes are mentioned for version 1607 in https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamew.

File: test/dshell/test20810.d [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module test.dshell.test20810;
+
+import dshell;
+
+int main()
+{
+    run("$DMD");  // should print usage and exit with 0
```
Comment (MoonlightSentinel) [inline test/dshell/test20810.d:7]: `dmd` without arguments returns exit code 1...
```suggestion
    run("$DMD --help");  // should print usage and exit with 0
```

**Comment by MoonlightSentinel** [comment]
You could also try a `compilable` test, e.g.:
```
/*
REQUIRED_ARGS: --help
PERMUTE_ARGS:
TEST_OUTPUT:
----
DMD64 D Compiler $r:.*$
Copyright (C) 1999-$n$ by The D Language Foundation, All Rights Reserved written by Walter Bright

Documentation: https://dlang.org/
Config file: 
Usage:
  dmd [<option>...] <file>...
  dmd [<option>...] -run <file> [<arg>...]

Where:
  <file>           D source file
  <arg>            Argument to pass when running the resulting program

<option>:
  @<cmdfile>       read arguments from cmdfile
  -allinst          generate code for all template instantiations
  -betterC          omit generating some runtime information and helper functions
  -boundscheck=[on|safeonly|off]
                    bounds checks on, in @safe only, or off
  -c                compile only, do not link
  -check=[assert|bounds|in|invariant|out|switch][=[on|off]]
                    enable or disable specific checks
  -check=[h|help|?] list information on all available checks
  -checkaction=[D|C|halt|context]
                    behavior on assert/boundscheck/finalswitch failure
  -checkaction=[h|help|?]
                    list information on all available check actions
  -color            turn colored console output on
  -color=[on|off|auto]
                    force colored console output on or off, or only when not redirected (default)
  -conf=<filename>  use config file at filename
  -cov              do code coverage analysis
  -cov=<nnn>        require at least nnn% code coverage
  -D                generate documentation
  -Dd<directory>    write documentation file to directory
  -Df<filename>     write documentation file to filename
  -d                silently allow deprecated features and symbols
  -de               issue an error when deprecated features or symbols are used (halt compilation)
  -dw               issue a message when deprecated features or symbols are used (default)
  -debug            compile in debug code
  -debug=<level>    compile in debug code <= level
  -debug=<ident>    compile in debug code identified by ident
  -debuglib=<name>  set symbolic debug library to name
  -defaultlib=<name>
                    set default library to name
  -deps             print module dependencies (imports/file/version/debug/lib)
  -deps=<filename>  write module dependencies to filename (only imports)
  -extern-std=<standard>
                    set C++ name mangling compatibility with <standard>
  -extern-std=[h|help|?]
                    list all supported standards
  -fPIC             generate position independent code
  -g                add symbolic debug info
  -gf               emit debug info for all referenced types
  -gs               always emit stack frame
  -gx               add stack stomp code
  -H                generate 'header' file
  -Hd=<directory>   write 'header' file to directory
  -Hf=<filename>    write 'header' file to filename
  -HC               generate C++ 'header' file
  -HCd=<directory>  write C++ 'header' file to directory
  -HCf=<filename>   write C++ 'header' file to filename
  --help            print help and exit
  -I=<directory>    look for imports also in directory
  -i[=<pattern>]    include imported modules in the compilation
  -ignore           ignore unsupported pragmas
  -inline           do function inlining
  -J=<directory>    look for string imports also in directory
  -L=<linkerflag>   pass linkerflag to link
  -lib              generate library rather than object files
  -lowmem           enable garbage collection for the compiler
  -m32              generate 32 bit code
  -m64              generate 64 bit code
  -main             add default main() (e.g. for unittesting)
  -man              open web browser on manual page
  -map              generate linker .map file
  -mcpu=<id>        generate instructions for architecture identified by 'id'
  -mcpu=[h|help|?]  list all architecture options
  -mixin=<filename> expand and save mixins to file specified by <filename>
  -mv=<package.module>=<filespec>
                    use <filespec> as source file for <package.module>
  -noboundscheck    no array bounds checking (deprecated, use -boundscheck=off)
  -O                optimize
  -o-               do not write object file
  -od=<directory>   write object & library files to directory
  -of=<filename>    name output file to filename
  -op               preserve source path for output files
  -preview=<id>     enable an upcoming language change identified by 'id'
  -preview=[h|help|?]
                    list all upcoming language changes
  -profile          profile runtime performance of generated code
  -profile=gc       profile runtime allocations
  -release          compile release version
  -revert=<id>      revert language change identified by 'id'
  -revert=[h|help|?]
                    list all revertable language changes
  -run <srcfile>    compile, link, and run the program srcfile
  -shared           generate shared library (DLL)
  -transition=<id>  help with language change identified by 'id'
  -transition=[h|help|?]
                    list all language changes
  -unittest         compile in unit tests
  -v                verbose
  -vcolumns         print character (column) numbers in diagnostics
  -verror-style=[digitalmars|gnu]
                    set the style for file/line number annotations on compiler messages
  -verrors=<num>    limit the number of error messages (0 means unlimited)
  -verrors=context  show error messages with the context of the erroring source line
  -verrors=spec     show errors from speculative compiles such as __traits(compiles,...)
  --version         print compiler version and exit
  -version=<level>  compile in version code >= level
  -version=<ident>  compile in version code identified by ident
  -vgc              list all gc allocations including hidden ones
  -vtls             list all variables going into thread local storage
  -w                warnings as errors (compilation will halt)
  -wi               warnings as messages (compilation will continue)
  -X                generate JSON file
  -Xf=<filename>    write JSON file to filename
  -Xcc=<driverflag> pass driverflag to linker driver (cc)
----
*/
```

**Comment by kinke** [comment]
> I'm not sure how this was never caught before though. But on my Win10 machine master fails.

I've just refactored this and misread the docs, that's why. And canonicalName() is either unused during the whole testsuite, or no tested DMD built with enabled assertions.

Anyway, I think a little unittest would be the best way to test this.

**Comment by AndrejMitrovic** [comment]
I thought `$r:.*$` would do substitution, but it doesn't.

And actually, I don't think this compilable test-case is correct. Because it's not testing the call properly. The `sc.ini` config path is empty here..

**Comment by MoonlightSentinel** [comment]
> I thought `$r:.*$` would do substitution, but it doesn't.

It does, the problem lies in the platform specific options (and a whitespace error due to markdown).

> And actually, I don't think this compilable test-case is correct. Because it's not testing the call properly. The `sc.ini` config path is empty here..

Actually, you're right because the test runner invokes dmd with `-conf`. Adding a unittest as @kinke 
 suggested should do the job.

Sorry for the distraction.

**Comment by AndrejMitrovic** [comment]
Alright, I'll fix it up later. 

> Sorry for the distraction.

Not at all. Thanks for help from both of you!

File: test/compilable/test20810.d [hunk 3]
```diff
@@ -0,0 +1,126 @@
+/*
+REQUIRED_ARGS: --help
```
Comment (NilsLankila) [inline test/compilable/test20810.d:2]: don't use `--help` but pass a simple module containing an empty main func. The use of `--help` is, in addition of not testing the right thing, something that would have to be maintained over years.

**Review by ghost — CHANGES_REQUESTED** [review]

File: test/compilable/test20810.d (hunk 3 above)
Comment (MoonlightSentinel) [reply test/compilable/test20810.d:2]: > don't use `--help` but pass a simple module containing an empty main func. The use of `--help` is, in addition of not testing the right thing.

Not an ideal test but running `dmd --help` triggers that bug.

> something that would have to be maintained over years.

Only the current version, I'll re-enable a maintenance free version after this PR is merged.

**Comment by MoonlightSentinel** [comment]
Any updates @AndrejMitrovic ?

File: test/compilable/test20810.d (hunk 3 above)
Comment (kinke) [reply test/compilable/test20810.d:2]: A brittle integration test for this is overkill; a trivial little unittest would do the job perfectly, just like https://github.com/dlang/dmd/blob/77ae1622340762d776fb5548b8cfbe27a6abd593/src/dmd/root/filename.d#L363-L370

Comment (MoonlightSentinel) [reply test/compilable/test20810.d:2]: Agreed, a unittest is the best solution here.

**Comment by AndrejMitrovic** [comment]
Well it seems that master works with your test, so it looks like the issue might be somewhere else?

**Comment by MoonlightSentinel** [comment]
it fails with current master, the PR was accidentally merged by dlang-bot and got reverted.

**Comment by AndrejMitrovic** [comment]
I'll close this since I can't reproduce it anymore, sorry.

**Pull request closed without merging**